consensus-engine/src/overlay/branch_overlay.rs

```diff
@@ -0,0 +1,294 @@
+use super::LeaderSelection;
+use crate::overlay::CommitteeMembership;
+use crate::{Committee, CommitteeId, NodeId, Overlay};
+use std::collections::HashMap;
+
+#[derive(Debug, Clone)]
+pub struct BranchOverlaySettings<L: LeaderSelection, M: CommitteeMembership> {
+    pub nodes: Vec<NodeId>,
+    pub current_leader: NodeId,
+    pub number_of_levels: usize,
+    pub leader: L,
+    pub committee_membership: M,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+/// Branch overlay with a single committee and round robin leader selection.
+pub struct BranchOverlay<L: LeaderSelection, M: CommitteeMembership> {
+    nodes: Vec<NodeId>,
+    leader: L,
+    committee_membership: M,
+    current_leader: NodeId,
+    number_of_committees: usize,
+    membership_committees: HashMap<usize, Committee>,
+    committees_by_member: HashMap<NodeId, usize>,
+}
+
+impl<L, M> Overlay for BranchOverlay<L, M>
+where
+    L: LeaderSelection + Send + Sync + 'static,
+    M: CommitteeMembership + Send + Sync + 'static,
+{
+    type Settings = BranchOverlaySettings<L, M>;
+    type LeaderSelection = L;
+    type CommitteeMembership = M;
+
+    fn new(settings: Self::Settings) -> Self {
+        let BranchOverlaySettings {
+            nodes,
+            current_leader,
+            number_of_levels,
+            leader,
+            committee_membership,
+        } = settings;
+        let (inner_committees, membership_committees) =
+            build_committee_from_nodes_with_size(&nodes, number_of_levels);
+
+        assert!(number_of_levels == inner_committees.len());
+
+        let committees_by_member = membership_committees
+            .iter()
+            .flat_map(|(committee, members)| members.iter().map(|member| (*member, *committee)))
+            .collect();
+        Self {
+            number_of_committees: number_of_levels,
+            nodes,
+            current_leader,
+            leader,
+            committee_membership,
+            membership_committees,
+            committees_by_member,
+        }
+    }
+
+    fn root_committee(&self) -> Committee {
+        self.membership_committees[&0].clone()
+    }
+
+    fn rebuild(&mut self, _timeout_qc: crate::TimeoutQc) {
+        // do nothing for now
+    }
```

Comment:
I think this needs to go out. If we didn't merge the refactor from other PR it will be done in that one directly.

Reply:
When #298 lands, I'll update this part, because `rebuild` is part of overlay trait.